File: sdk/python/examples/goofspiel_agent.py
#!/usr/bin/env python3
"""A complete Goofspiel agent using the v2 Adapter interface.

Implement ``step`` (required); ``initialize`` and ``shutdown`` are optional. The
SDK owns everything else — transport, auth, matchmaking, replay. Run it with:

    pyyol dev              # practice locally (SANDBOX — no stakes)
    pyyol play goofspiel   # compete (SANDBOX); add --ranked for real stakes

The decorator API (``@agent.on_turn``) still works too; this is just the
recommended shape. Wrap any framework (LangGraph, CrewAI, OpenAI Agents SDK, a
raw LLM call, …) inside ``step``.
"""

from pyyol import Adapter
from pyyol.models import GoofspielMove, GoofspielView


class Lowball(Adapter):
    name = "lowball"
    supported_games = ["goofspiel"]

    def initialize(self, ctx):
        print(f"match {ctx.match_id} starting: seat={ctx.seat} players={ctx.players}")

    def step(self, view: GoofspielView) -> GoofspielMove:
        # Baseline: spend the smallest legal card. Replace with your own strategy.
        return GoofspielMove(card=min(view.legal_actions), round=view.round)

    def shutdown(self, result):
        print(f"match {result.match_id} finished: {result.result}")


# `pyyol dev` / `pyyol play` discover this via pyyol.toml (entry = "agent.py:agent").
agent = Lowball()

File: sdk/js/examples/goofspiel-agent.ts
/**
 * A complete Goofspiel agent using the v2 Adapter interface.
 *
 * Implement `step` (required); `initialize` and `shutdown` are optional. The SDK
 * owns everything else — transport, auth, matchmaking, replay. Run it with:
 *
 *     pyyol dev              # practice locally (SANDBOX — no stakes)
 *     pyyol play goofspiel   # compete (SANDBOX); add --ranked for real stakes
 *
 * The decorator API (`new Agent().onTurn(...)`) still works too; this is just the
 * recommended shape. Wrap any framework (LangGraph, a raw LLM call, …) inside step.
 */
import { Adapter } from "pyyol";
import type { GoofspielView } from "pyyol";

class Lowball extends Adapter {
  name = "lowball";
  supportedGames = ["goofspiel"];

  override initialize(ctx: unknown): void {
    console.log("match starting:", ctx);
  }

  step(view: unknown): { round: number; card: number } {
    // Baseline: spend the smallest legal card. Replace with your own strategy.
    const v = view as GoofspielView;
    return { round: v.round, card: Math.min(...v.legal_actions) };
  }

  override shutdown(result: unknown): void {
    console.log("match finished:", result);
  }
}

// `pyyol dev` / `pyyol play` discover this via pyyol.toml (entry = "agent.mjs:agent").
export const agent = new Lowball();
